src/v/kafka/server/handlers/metadata.cc

```diff
@@ -422,4 +423,31 @@ ss::future<response_ptr> metadata_handler::handle(
     co_return co_await ctx.respond(std::move(reply));
 }
 
+size_t metadata_memory_estimator(size_t) {
+    // We cannot make a precise estimate of the size of a metadata response by
+    // examining only the size of the request (nor even by examining the entire
+    // request) since the response depends on the number of partitions in the
+    // cluster. Instead, we return a conservative estimate based on the soft
+    // limit of total cluster-wide partition counts, mutiplied by an empirical
+    // "bytes per partition".
+    //
+    // The bytes per partition is roughly split evenly between the number of
+    // bytes needed to encode the response (about 100 bytes per partition,
+    // assuming 3 replicas) as well as the input structure to the encoder (i.e.,
+    // the in-memory representation of that same metadata).
+    //
+    // This estimate is additionally inexact in the sense that more a higher
+    // replication factor would increase the per-partition size without bound.
+    // An additional inaccuracy is that we don't consider topics but lump all
+    // partitions together: this works for moderate to large partition counts,
+    // but it likely that a cluster with (for example) many topics with only 1
+    // partition each might produce a larger metadata response than this
+    // calculation accounts for.
+
+    // Empirical highwater memory allocated per partition while processing a
+    // metadata response.
+    constexpr size_t bytes_per_partition = 200;
+
+    return max_clusterwide_partitions * bytes_per_partition;
```

Comment:
"Probably" - but I explicitly wanted to avoid this. This early request processing should be as simple as possible, so use information we have statically available which is just the request size at this point.

At least in the scenario where we saw this problem, the topic metadata wasn't necessarily populated yet because the system was just starting up so in some cases you wouldn't have a reliable estimate, though I guess you could use the conservative estimate at that point, but in general this seems subject to data staleness (among others, the OOM was seen in other cases too e.g., where the controller was killed and so requests were piling up waiting for a new controller to be elected: in this case the old topic table would probably be good enough).

If this were the final change, I would perhaps be in favor of this: but it is only a stopgap towards what I think should be the final solution: a two phase allocation of units where we allocate a small number of units up front to account for initial processing, then the "full amount" at the point during processing that we know the final amount, e.g., once we've queried the topic table and before we start serializing the result. I'll raise this solution up for more discussion, however.

Reply:
> That sounds strange: we don't start listening on the Kafka port until the controller

The server was up and accepting kafka connections, but the metadata requests were blocked on refreshing metadata from the controller (which is what caused requests to line up in time). As I recall there had never been a successful metadata dissemination to this node at the time: by the time the dissemination completed the requests were already long past the initial part of the connection and were blocked on the metadata refresh semaphore.

Regardless of the specific details of that case, doing the estimation separately is always subject to races? You do the estimation and then some time later prepare the response and the number of partitions may have changed arbitrarily in the meantime.

> I was thinking of the two phase allocation the other way around: that we would initially allocate the worst case number of units, and then release some when we knew enough about the request to do so.

Yes, I recall. The primary complaint (not from me) was the possible performance impact due to reduced concurrency. Though see below also.

> If we do it the other way (small allocation then big allocation), we risk deadlock (or just reduced throughput) from too many concurrent requests taking their small phase 1 units, then there not being enough units available to allow any of them to proceed into phase 2.

Yes, I am aware. I don't know if this is the right place for a detailed review of this final state, but I do consider this case by including a (1) portion of the units which are guaranteed to be only for phase 2 and (2) prioritization of the units semaphore so in-progress requests always get first shots at the units. It was the realization that we need to do (2) that made me want to do this stopgap first before proceeding with the full fix.

> So I guess the key question here is whether the end state is going to be the small->big procedure

My primary complaint about big -> small is that I don't think it is very general. It _happens_ to work for a specific type of metadata request (the "empty" request which means "give me everything") which has a known max size and where the size is fairly easily to calculate based on cached metadata, but in general you might have to a lot more work to estimate the size: e.g., any metadata request which a topic list (though an overestimate would at least work) or other types of requests. At this point we don't even have access to the body of the request (though we could change that): we've read only 4 bytes (the size). 

In the limit, you have to more or less do the request twice: once to estimate the size and then once again to really do it. The small -> big just kind of works "naturally" in that you don't have a separate estimation phase, you'll always have some point where you know the size of the request as you are processing it (at least, one can hope, though there is the question of whether you know it _before_ you use the memory, which is usually possible but often not the way it is currently written).

If it were _only_ this one type of metadata request, I think this would be fine but this is really a problem with a lot of request types, which I think favors small -> big if we can convince ourselves of deadlock freedom. Note that small -> big can really be something like small -> big -> small too, e.g., we can free units after we leave the high memory part of the processing such as in the other comment (though to avoid deadlock there should only be 1 increases after the initial one).

Reply:
Though I wonder if we can decouple the "full fix" discussion from the "stopgap one"? Whichever way we go I don't think this change really pre-disposes us either way. Most of the changes are just related to allowing per-handler estimates _at all_ and neither method of plumbing the units around is implemented here (well really, big -> small doesn't need any plumbing of units, it just needs a richer estimator, like a second `handle` type call to do the estimation, while small -> big needs the units put into the request context so they can be adjusted).

Reply:
> As I recall there had never been a successful metadata dissemination to this node at the time so as far as the node was concerned there were ~0 partitions, 

Ah, so this sounds like it could have been avoided by consulting topics_table (populated before kafka api comes up by controller replay) rather than any of the metadata dissemination state which is fuzzier.

Anyway.  I guess this boils down to "is hardcoded 40k good enough for 22.3".  I think what I'm missing is how hard it would actually be to wire through topics_table: if it's super onerous then I can accept hardcoding a magic number, but can you maybe try it and see how invasive the patch is?  and/or for concerns about races, we could have a hardcoded lower bound that says "even if you didn't see partitions yet, pretend like there are at least 10k" or somesuch.

Maybe a compromise is to make it a config property rather than hardcoding it -- I was thinking of having one of these anyway, to encode the "here's how big we ever tested" number.  That way folks doing scale testing beyond 40k can at least dial it up without recompiling redpanda.

Reply:
> Ah, so this sounds like it could have been avoided by consulting topics_table (populated before kafka api comes up by controller replay) rather than any of the metadata dissemination state which is fuzzier.

I believe this was the first startup _ever_ of the node so I suspect the controller state is at least initially empty, though maybe you said above that on first join we don't start accepting connections until controller state has been recovered/replayed from another node?

> Anyway. I guess this boils down to "is hardcoded 40k good enough for 22.2". I think what I'm missing is how hard it would actually be to wire through topics_table: if it's super onerous then I can accept hardcoding a magic number, but can you maybe try it and see how invasive the patch is? 

This should be fairly easy and yes I can do that.

> we could have a hardcoded lower bound that says "even if you didn't see partitions yet, pretend like there are at least 10k" or somesuch.

Yes, though the race isn't only for the no-partitions case, it can be at any point: from jumping from 20k to 100k partitions or whatever. I think this a downside of big -> small in general: any estimate can always become out of date between the estimation point and when the request is processed, and even though this window might be small with constant requests we would expect it to be hit by some requests with high probability when the conditions change (though a fix for this could be to fail requests that are found to be unexpectedly large rather than forging ahead into a possible OOM).

> Maybe a compromise is to make it a config property rather than hardcoding it -- I was thinking of having one of these anyway, to encode the "here's how big we ever tested" number. That way folks doing scale testing beyond 40k can at least dial it up without recompiling redpanda.

Also a good suggestion though I'll try your other suggestion first.

Reply:
I have updated this PR to use an estimate based on the actual topic metadata.